Update conda instructions

saraedum opened the pull request:
The conda instructions are a bit more complicated than they have to be. They also mention some things that are deprecated (Mambaforge) or uncessary (mamba). 

Here, we try to simplify the instructions with the experience we had by running installations on SageDays 128.

<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion. Nothing that I would know of.
- [x] I have created tests covering the changes. No, it's quite complicated to test these things, i.e., making sure that the actual instructions work and the tests stay in sync with the instructions.
- [x] I have updated the documentation and checked the documentation preview.

saraedum committed:
Do not recommend mambaforge, it is deprecated

saraedum committed:
Do not recommend mamba

mamba is great, and it's sometimes a bit faster than conda still. But
conda has finally upgraded their solver so the conda solver works
perfectly fine for our purposes.

saraedum committed:
No need for SAGE_NUM_THREADS

it's figured out correctly automatically on Linux on macOS (tested at
SD128.)

saraedum committed:
Simplify conda installation instructions and normalize layout

saraedum committed:
Clarify the channel setup instructions

most people reading this probably have no idea what channels are.

saraedum commented on src/doc/en/installation/conda.rst:
```diff
@@ -73,51 +57,36 @@ environment for Sage development.
-
-    .. code-block:: shell
+  and a few additional developer tools; if you are not on the most common types
+  of Linux, replace ``linux`` accordingly:
```
This might seem obvious but it was not to a participant of sd128.

vincentmacri replied:
I'm not sure what is meant by "the most common types of Linux", could you clarify it in the instructions?

saraedum replied:
Ok, I was hoping that it would be less confusing this way.  I'll make the statement more technical.

saraedum replied:
```suggestion
  and a few additional developer tools; if you are not on a Linux that runs an x86_64 compatible CPU,
  replace ``linux`` accordingly:
```

saraedum commented on src/doc/en/installation/conda.rst:
```diff
@@ -156,5 +131,5 @@ To build the documentation, use::
   You can update the conda lock files by running
   ``.github/workflows/conda-lock-update.py`` or by running
-  ``conda-lock --platform linux-64 --filename environment-3.11-linux.yml --lockfile environment-3.11-linux.lock``
+  ``conda-lock --platform linux-64 --filename environment-3.12-linux.yml --lockfile environment-3.12-linux.lock``
```
I updated some strings in this file. Not that it really matters but maybe it is going to keep the commands valid literally for a little bit longer.

saraedum committed:
Explain how to enable JupyterLab in this setup

vincentmacri commented on src/doc/en/installation/conda.rst:
```diff
@@ -133,16 +102,29 @@ suffices to restart Sage.
+
+    $ conda env update --file environment-3.12-linux.yml
+
+If you want to use Jupyter notebooks, you need to install JupyterLab and the SageMath kernel::
```
I'm not sure if it's possible to do this, but if it is could you also explain how to create a Jupyter kernel for use in an existing/system Jupyter installation? This is something I've had trouble doing in conda so if there's a way to do it it would be nice to document.

saraedum replied:
That's more complicated. You need to wrap things into an activation script and it's going to end up being more brittle. I am not sure whether I would want to advertise this here. (And I also don't want to test whether these instructions really work on Linux/macOS or even WSL.)

vincentmacri replied:
Fair enough.

saraedum committed:
Clarify what the usual linux flavors are in the conda guide

saraedum committed:
Fix RST errors in conda documentation

user202729 commented on src/doc/en/installation/conda.rst:
```diff
@@ -9,48 +9,32 @@ SageMath can be installed on Linux and macOS via Conda from the
+.. code-block:: shell
 
-where ``X`` is version of Python, e.g. ``3.9``.
+    $ conda create -n sage sage
```
If you only keep one, I recommend using mamba. It's drop-in replacement and much faster.

saraedum replied:
I personally use mamba but conda works just fine and it is available on the setups that we are covering here (see the top) i.e., also on miniconda and anaconda.

vincentmacri replied:
I agree with maintaining conda instructions as I think it's a bit more widely available in various distro repositories. No opinion on keeping both or having just conda.

user202729 commented on src/doc/en/installation/conda.rst:
```diff
@@ -73,51 +57,36 @@ environment for Sage development.
-  whatever the meaningful value for your machine is - no more than
-  the number of cores::
-
-    $ export SAGE_NUM_THREADS=24
```
Instead of removing optional stuff entirely I recommend moving it down to "notes" section. It might be useful sometimes e.g. I only want to use 3 CPUs for compilation to avoid the computer getting too slow doing something else.

user202729 commented on src/doc/en/installation/conda.rst:
```diff
@@ -73,51 +57,36 @@ environment for Sage development.
 
 Here we assume that you are using a git checkout.
 
-- Optionally, set the build parallelism for the Sage library. Use
```
think about it, why does [this section](https://doc-release--sagemath.netlify.app/html/en/installation/conda#using-conda-to-provide-all-dependencies-for-the-sage-library) even exist? isn't https://doc-release--sagemath.netlify.app/html/en/installation/meson enough?

saraedum replied:
Yes, eventually this section should be removed. But I understood that the meson setup isn't sufficiently stable and some people are using the setup described here.

@tobiasdiez @dimpase do you think that the meson setup should be recommended for everybody that uses conda dependencies and we should just remove this page altogether?

vincentmacri replied:
It's been a few months since I tried using the meson setup but I remember having issues with it.

tobiasdiez replied:
Yes, in my opinion the meson build is stable enough to be the default for conda-based setups. So I would welcome if we would merge the meson instructions with the conda ones, but would of course understand if you don't want to do this as part of this PR.

saraedum replied:
@vincentmacri would you have the time to try if the meson setup works for you now? I just tried it and it worked fine for me. I'd prefer to delete a page from the documentation if possible.

vincentmacri replied:
Meson is still giving me issues. It's possible the problem is on my end, but I'm following the steps in the documentation so if something is wrong on my end I think the meson documentation needs to be updated to be more clear.

On Fedora 41, from the root of the develop branch of my sage fork which has no changes from the develop branch in the sagemath repo:

```
conda env create --file environment-3.11-linux.yml --name sage-meson
conda activate sage-meson
pip install --no-build-isolation --editable .
./sage
```

gives the error:

```
Error in sys.excepthook:
Traceback (most recent call last):
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/pathlib.py", line 1250, in is_dir
    return S_ISDIR(self.stat().st_mode)
                   ^^^^^^^^^
AttributeError: 'str' object has no attribute 'stat'

Original exception was:
Traceback (most recent call last):
  File "/home/vince/Documents/git/sage/src/bin/sage-ipython", line 15, in <module>
    app.initialize()
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/site-packages/traitlets/config/application.py", line 118, in inner
    return method(app, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/site-packages/IPython/terminal/ipapp.py", line 278, in initialize
    self.init_shell()
  File "/home/vince/Documents/git/sage/src/sage/repl/interpreter.py", line 854, in init_shell
    self.shell.extension_manager.load_extension(SAGE_EXTENSION)
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/site-packages/IPython/core/extensions.py", line 62, in load_extension
    return self._load_extension(module_str)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/site-packages/IPython/core/extensions.py", line 79, in _load_extension
    if self._call_load_ipython_extension(mod):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vince/.conda/envs/sage-dev/lib/python3.11/site-packages/IPython/core/extensions.py", line 129, in _call_load_ipython_extension
    mod.load_ipython_extension(self.shell)
  File "/home/vince/Documents/git/sage/src/sage/repl/__init__.py", line 6, in load_ipython_extension
    sage.repl.ipython_extension.load_ipython_extension(*args)
  File "/home/vince/Documents/git/sage/src/sage/misc/misc.py", line 1029, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/vince/Documents/git/sage/src/sage/repl/ipython_extension.py", line 744, in load_ipython_extension
    SageCustomizations(shell=ip)
  File "/home/vince/Documents/git/sage/src/sage/repl/ipython_extension.py", line 591, in __init__
    import sage.all  # noqa: F401
    ^^^^^^^^^^^^^^^
  File "/home/vince/Documents/git/sage/src/sage/all.py", line 97, in <module>
    from sage.graphs.all import *
  File "/home/vince/Documents/git/sage/src/sage/graphs/all.py", line 9, in <module>
    from sage.graphs.graph import Graph
  File "/home/vince/Documents/git/sage/src/sage/graphs/graph.py", line 436, in <module>
    class Graph(GenericGraph):
  File "/home/vince/Documents/git/sage/src/sage/graphs/graph.py", line 9273, in Graph
    from sage.graphs.connectivity import minimal_separators
ImportError: cannot import name 'minimal_separators' from 'sage.graphs.connectivity' (/home/vince/Documents/git/sage/src/sage/graphs/connectivity.cpython-311-x86_64-linux-gnu.so)
111

user202729 replied:
`/home/vince/Documents/git/sage/src/sage/repl/ipython_extension.py`

Implies your sage git repository root is `/home/vince/Documents/git/sage/`.

`ImportError: cannot import name 'minimal_separators' from 'sage.graphs.connectivity' (/home/vince/Documents/git/sage/src/sage/graphs/connectivity.cpython-311-x86_64-linux-gnu.so)`

implies the `.so` file is found right next to the `.pyx` file.

This should not normally happen, normally meson builds out of source.

Maybe you need to do the following

> `If you have previously build Sage in-place, you first have to delete the already compiled files, e.g. with shopt -s globstar followed by rm src/**/*.so or for f in src/**/*.so ; do mv "$f" "$f.old"; done. Moreover, remove the old generated files with find src/sage/ext/interpreters -type f ! -name 'meson.build' -delete. Also uninstall the ‘old’ sage packages with pip uninstall sage-conf sage-setup sagemath-standard.`

That said, I find `git clean -fxd` easier (though it might be more brutal because it may remove some user-generated files that is already gitignored)

vincentmacri replied:
I had tried what the doc mentioned to remove old generated files before I posted that error. After doing `git clean -fxd` though everything worked. I guess the meson conda build works for me now. Thanks!

user202729 replied:
That's weird (maybe your shell doesn't support globstar? Are you using something that is not bash?)

vincentmacri replied:
> That's weird (maybe your shell doesn't support globstar? Are you using something that is not bash?)

Nope, I'm using bash.

saraedum committed:
Fix formatting of texts after code blocks

saraedum committed:
Drop bullets that do not help readability

saraedum committed:
Put SAGE_NUM_THREADS back into the conda instructions

saraedum committed:
Simplify confusing statement about pip install

while that statement might be technically true somehow, I think for most
people this just "builds" Sage. That it is "installed" somehow is maybe
misleading.
